File: backend/fibda/transfers.py
"""Bounded, non-executing imports and image normalization."""
import csv
import io
import warnings
import zipfile
from pathlib import PurePosixPath

MAX_FILE = 20 * 1024 * 1024
MAX_TOTAL = 200 * 1024 * 1024
MAX_ROWS = 10000
FIELDS = {'first_name','last_name','birth_date','sex','nationalities','country','club','section','height_cm','weight_kg','category_id'}


def safe_members(archive, max_total=MAX_TOTAL):
    infos = archive.infolist()
    if len(infos) > 2000 or sum(x.file_size for x in infos) > max_total:
        raise ValueError('Archive trop volumineuse')
    seen = set()
    for info in infos:
        path = PurePosixPath(info.filename)
        if (path.is_absolute() or '..' in path.parts or '\\' in info.filename or ':' in info.filename
                or info.filename in seen or ((info.external_attr >> 16) & 0o170000) == 0o120000):
            raise ValueError('Chemin dangereux ou doublon dans archive')
        seen.add(info.filename)
        if info.file_size > MAX_FILE:
            raise ValueError('Fichier trop volumineux')
    return infos


def preview_import(data, filename):
    if len(data) > MAX_FILE:
        raise ValueError('Fichier trop volumineux')
    errors, rows, notices = [], [], []
    if filename.lower().endswith('.csv'):
        text = data.decode('utf-8-sig')
        try:
            dialect = csv.Sniffer().sniff(text[:8192], delimiters=',;\t')
        except csv.Error:
            dialect = csv.excel
        raw = list(csv.reader(io.StringIO(text), dialect))
    elif filename.lower().endswith('.xlsx'):
        from openpyxl import load_workbook
        with zipfile.ZipFile(io.BytesIO(data)) as z:
            safe_members(z)
        workbook = load_workbook(io.BytesIO(data), read_only=True, data_only=False)
        raw = []
        for index, row in enumerate(workbook.active.iter_rows(values_only=True)):
            if index > MAX_ROWS:
                raise ValueError('Maximum 10000 lignes')
            from datetime import date, datetime
            raw.append([value.date().isoformat() if isinstance(value, datetime) else value.isoformat() if isinstance(value, date) else value for value in row])
        workbook.close()
    else:
        raise ValueError('Formats acceptés : CSV UTF-8 et XLSX')
    if not raw:
        return {'rows': [], 'errors': [{'row': 1, 'message': 'Fichier vide'}], 'warnings': []}
    headers = [str(x or '').strip() for x in raw[0]]
    required = {'first_name','last_name','birth_date','sex','section','country','nationalities'}
    if len(headers) != len(set(headers)) or not required.issubset(headers):
        errors.append({'row': 1, 'message': 'Entêtes uniques obligatoires : first_name, last_name, birth_date, sex, section, country, nationalities'})
    unknown = set(headers) - FIELDS
    if unknown:
        errors.append({'row': 1, 'message': 'Colonnes inconnues : ' + ', '.join(sorted(unknown))})
    if len(raw) > MAX_ROWS + 1:
        raise ValueError('Maximum 10000 lignes')
    seen = set()
    for n, values in enumerate(raw[1:], 2):
        if not any(x is not None and str(x).strip() for x in values):
            continue
        row = {k: str(v or '').strip() for k, v in zip(headers, values)}
        if len(values) != len(headers):
            errors.append({'row': n, 'message': 'Nombre de colonnes incohérent'})
        if any(str(v).lstrip().startswith(('=', '+', '-', '@')) for v in values if v is not None):
            errors.append({'row': n, 'message': 'Formule ou préfixe exécutable refusé'})
        if not row.get('first_name') or not row.get('last_name'):
            errors.append({'row': n, 'message': 'Nom et prénom obligatoires'})
        if row.get('sex') not in ('M', 'F'):
            errors.append({'row': n, 'message': 'Sexe attendu M ou F'})
        if row.get('section') not in ('amateur', 'pro'):
            errors.append({'row': n, 'message': 'Section attendue amateur ou pro'})
        for key in ('height_cm', 'weight_kg'):
            if row.get(key):
                try:
                    from decimal import Decimal, InvalidOperation
                    value = Decimal(row[key].replace(',', '.'))
                    if not value.is_finite() or value <= 0:
                        raise InvalidOperation
                    row[key] = str(value)
                except InvalidOperation:
                    errors.append({'row': n, 'message': key + ' doit être un nombre positif'})
        if not row.get('birth_date'):
            errors.append({'row': n, 'message': 'Date de naissance obligatoire'})
        if row.get('birth_date'):
            try:
                from datetime import date
                date.fromisoformat(row['birth_date'])
            except ValueError:
                errors.append({'row': n, 'message': 'Date attendue AAAA-MM-JJ'})
        key = (row.get('first_name', '').casefold(), row.get('last_name', '').casefold(), row.get('birth_date'))
        if key in seen:
            notices.append({'row': n, 'message': 'Identité répétée : vérifier avant import'})
        seen.add(key)
        if row.get('nationalities'):
            row['nationalities'] = [x.strip().upper() for x in row['nationalities'].split('|') if x.strip()]
        countries = [row.get('country', '')] + row.get('nationalities', []) if isinstance(row.get('nationalities'), list) else []
        if len(countries) < 2 or any(len(c) != 2 or not c.isascii() or not c.isalpha() or c != c.upper() for c in countries):
            errors.append({'row': n, 'message': 'Pays et nationalités obligatoires : codes de deux lettres majuscules ; séparer les nationalités par |'})
        rows.append(row)
    return {'rows': rows, 'errors': errors, 'warnings': notices}


def sanitize_photo(data, crop=None):
    from PIL import Image, ImageOps
    if len(data) > MAX_FILE:
        raise ValueError('Photo trop volumineuse')
    with warnings.catch_warnings():
        warnings.simplefilter('error', Image.DecompressionBombWarning)
        try:
            with Image.open(io.BytesIO(data)) as source:
                if source.format not in ('JPEG', 'PNG', 'WEBP') or source.width * source.height > 25000000:
                    raise ValueError('Photo non supportée ou dimensions excessives')
                source.load()
                picture = ImageOps.exif_transpose(source).convert('RGB')
        except (OSError, Image.DecompressionBombError, Image.DecompressionBombWarning) as exc:
            raise ValueError('Photo invalide') from exc
    if crop is not None:
        if len(crop) != 4 or not all(isinstance(x, (int, float)) for x in crop):
            raise ValueError('Recadrage attendu [gauche,haut,droite,bas] en pixels')
        left, top, right, bottom = crop
        if not (0 <= left < right <= picture.width and 0 <= top < bottom <= picture.height):
            raise ValueError('Recadrage hors photo')
        picture = picture.crop(crop)
    picture.thumbnail((1600, 2000))
    output = io.BytesIO()
    picture.save(output, 'JPEG', quality=88, optimize=True)
    return output.getvalue()


def batch_photos(data, mappings):
    if len(data) > MAX_TOTAL:
        raise ValueError('Archive trop volumineuse')
    result, owners = [], set()
    with zipfile.ZipFile(io.BytesIO(data)) as archive:
        available = {x.filename for x in safe_members(archive) if not x.is_dir()}
        for mapping in mappings:
            name = mapping.get('filename')
            owner = (mapping.get('owner_type'), mapping.get('owner_id'), mapping.get('kind'))
            if name not in available or owner[0] not in ('person', 'official') or not owner[1] or owner[2] not in ('portrait', 'full') or owner in owners:
                raise ValueError('Association photo absente, invalide ou répétée')
            owners.add(owner)
            result.append(dict(owner_type=owner[0], owner_id=owner[1], kind=owner[2], data=sanitize_photo(archive.read(name), mapping.get('crop'))))
    return result

File: frontend/server/transfers.ts
import { Problem } from "./problem";
import { uid } from "./util";
import type { Store, Conn } from "./store";
import type { User } from "./state";
import { applyCommand } from "./commands";

// Portage de backend/fibda/transfers.py (partie imports) : contrôle borné, sans exécution,
// d'un fichier d'inscriptions avant son import. Mêmes messages et même forme de réponse que
// le Python ({rows, errors, warnings}), consommés tels quels par frontend/src/Preparation.tsx.
//
// Différences assumées avec le Python :
// - taille maximale 4 Mo (Vercel refuse un corps de requête au-delà de 4,5 Mo) au lieu de 20 Mo ;
// - XLSX indisponible : les bibliothèques npm candidates (exceljs 4.4.0, xlsx 0.18.5) ont des
//   vulnérabilités connues à l'audit (voir rapport de tâche), on répond 415 « convertir en CSV » ;
// - la prévisualisation est conservée en base (table import_previews), pas en mémoire, car deux
//   appels serverless peuvent tomber sur deux instances distinctes.

export const MAX_FILE = 4 * 1024 * 1024;
export const MAX_ROWS = 10000;
export const PREVIEW_SECONDS = 3600;
export const FIELDS = new Set(["first_name", "last_name", "birth_date", "sex", "nationalities", "country", "club", "section", "height_cm", "weight_kg", "category_id"]);
const REQUIRED = ["first_name", "last_name", "birth_date", "sex", "section", "country", "nationalities"];

export type ImportMessage = { row: number; message: string };
export type ImportPreview = { rows: Record<string, any>[]; errors: ImportMessage[]; warnings: ImportMessage[] };

// --- CSV : parseur maison strict, sans dépendance ---------------------------------------------

// Séparateur détecté sur la première ligne, hors guillemets : le plus fréquent parmi ; , et
// tabulation (comme le Sniffer Python) ; virgule par défaut.
export function detectDelimiter(text: string): string {
  const counts: Record<string, number> = { ",": 0, ";": 0, "\t": 0 };
  let quoted = false;
  for (const ch of text) {
    if (ch === '"') quoted = !quoted;
    else if (!quoted && (ch === "\n" || ch === "\r")) break;
    else if (!quoted && ch in counts) counts[ch] += 1;
  }
  let best = ",";
  for (const d of [";", "\t"]) if (counts[d] > counts[best]) best = d;
  return best;
}

// Découpe en lignes et cellules. Guillemets doublés à l'intérieur d'un champ entre guillemets,
// retours de ligne CRLF ou LF. Une ligne mal formée (guillemet ouvrant non fermé, guillemet au
// milieu d'un champ, texte après le guillemet fermant) est refusée avec son numéro.
export function parseCsv(text: string, delimiter?: string): string[][] {
  if (text.startsWith("\uFEFF")) text = text.slice(1);
  delimiter ??= detectDelimiter(text);
  const rows: string[][] = [];
  let row: string[] = [];
  let field = "";
  let quoted = false;
  let atFieldStart = true;
  let line = 1;
  // Ligne où le champ en cours a commencé : c'est elle qu'on signale, même si un champ entre
  // guillemets s'étend sur plusieurs lignes.
  let fieldLine = 1;
  const malformed = (at = fieldLine) => new Problem(`Ligne ${at} mal formée : guillemets incorrects`);
  const endRow = () => {
    row.push(field);
    // Ligne vide : conservée comme ligne sans cellule, ignorée ensuite comme en Python.
    rows.push(row.length === 1 && row[0] === "" ? [] : row);
    row = [];
    field = "";
    atFieldStart = true;
  };
  for (let i = 0; i < text.length; i++) {
    const ch = text[i];
    if (quoted) {
      if (ch === '"') {
        if (text[i + 1] === '"') {
          field += '"';
          i += 1;
        } else {
          quoted = false;
          // Après le guillemet fermant : séparateur ou fin de ligne uniquement.
          const next = text[i + 1];
          if (next !== undefined && next !== delimiter && next !== "\n" && next !== "\r") throw malformed();
        }
      } else {
        if (ch === "\n") line += 1;
        field += ch;
      }
      continue;
    }
    if (ch === '"') {
      if (!atFieldStart) throw malformed(line);
      quoted = true;
      fieldLine = line;
      atFieldStart = false;
    } else if (ch === delimiter) {
      row.push(field);
      field = "";
      atFieldStart = true;
    } else if (ch === "\r" || ch === "\n") {
      if (ch === "\r" && text[i + 1] === "\n") i += 1;
      endRow();
      line += 1;
    } else {
      field += ch;
      atFieldStart = false;
    }
  }
  if (quoted) throw malformed();
  if (field !== "" || row.length) endRow();
  return rows;
}

// --- Contrôles ligne à ligne : copie de preview_import -----------------------------------------

const asText = (v: unknown): string => (v === null || v === undefined ? "" : String(v));
const isFormula = (v: string): boolean => /^\s*[=+\-@]/.test(v);
const isCountry = (c: string): boolean => /^[A-Z]{2}$/.test(c);
// Nombre positif fini, comme Decimal(...) > 0 côté Python ; la virgule décimale est acceptée.
const NUMBER = /^[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?$/;
const isIsoDate = (value: string): boolean => {
  const m = /^(\d{4})-(\d{2})-(\d{2})$/.exec(value);
  if (!m) return false;
  const [y, mo, d] = [Number(m[1]), Number(m[2]), Number(m[3])];
  const date = new Date(Date.UTC(y, mo - 1, d));
  return date.getUTCFullYear() === y && date.getUTCMonth() === mo - 1 && date.getUTCDate() === d;
};

function decodeUtf8(data: Uint8Array): string {
  try {
    // « utf-8-sig » : le BOM est retiré, un octet invalide refuse le fichier.
    return new TextDecoder("utf-8", { fatal: true, ignoreBOM: false }).decode(data);
  } catch {
    throw new Problem("Fichier non lisible : encodage UTF-8 attendu");
  }
}

export function previewImport(data: Uint8Array, filename: string): ImportPreview {
  if (data.byteLength > MAX_FILE) throw new Problem("Fichier trop volumineux");
  const lower = filename.toLowerCase();
  let raw: unknown[][];
  if (lower.endsWith(".csv")) raw = parseCsv(decodeUtf8(data));
  else if (lower.endsWith(".xlsx")) throw new Problem("Format XLSX indisponible : convertir en CSV", 415);
  else throw new Problem("Formats acceptés : CSV UTF-8 et XLSX");
  const errors: ImportMessage[] = [];
  const warnings: ImportMessage[] = [];
  const rows: Record<string, any>[] = [];
  if (!raw.length) return { rows: [], errors: [{ row: 1, message: "Fichier vide" }], warnings: [] };
  const headers = raw[0].map((x) => asText(x).trim());
  if (new Set(headers).size !== headers.length || !REQUIRED.every((h) => headers.includes(h))) {
    errors.push({ row: 1, message: "Entêtes uniques obligatoires : first_name, last_name, birth_date, sex, section, country, nationalities" });
  }
  const unknown = headers.filter((h) => !FIELDS.has(h)).sort();
  if (unknown.length) errors.push({ row: 1, message: "Colonnes inconnues : " + unknown.join(", ") });
  if (raw.length > MAX_ROWS + 1) throw new Problem("Maximum 10000 lignes");
  const seen = new Set<string>();
  for (let index = 1; index < raw.length; index++) {
    const values = raw[index];
    const n = index + 1;
    if (!values.some((x) => x !== null && x !== undefined && asText(x).trim())) continue;
    const row: Record<string, any> = {};
    headers.forEach((k, i) => {
      if (i < values.length) row[k] = asText(values[i]).trim();
    });
    if (values.length !== headers.length) errors.push({ row: n, message: "Nombre de colonnes incohérent" });
    if (values.some((v) => v !== null && v !== undefined && isFormula(String(v)))) errors.push({ row: n, message: "Formule ou préfixe exécutable refusé" });
    if (!row.first_name || !row.last_name) errors.push({ row: n, message: "Nom et prénom obligatoires" });
    if (!["M", "F"].includes(row.sex)) errors.push({ row: n, message: "Sexe attendu M ou F" });
    if (!["amateur", "pro"].includes(row.section)) errors.push({ row: n, message: "Section attendue amateur ou pro" });
    for (const key of ["height_cm", "weight_kg"]) {
      if (row[key]) {
        const normalized = String(row[key]).replace(",", ".");
        const value = Number(normalized);
        if (!NUMBER.test(normalized) || !Number.isFinite(value) || value <= 0) errors.push({ row: n, message: key + " doit être un nombre positif" });
        else row[key] = normalized;
      }
    }
    if (!row.birth_date) errors.push({ row: n, message: "Date de naissance obligatoire" });
    else if (!isIsoDate(row.birth_date)) errors.push({ row: n, message: "Date attendue AAAA-MM-JJ" });
    const identity = JSON.stringify([(row.first_name ?? "").toLowerCase(), (row.last_name ?? "").toLowerCase(), row.birth_date]);
    if (seen.has(identity)) warnings.push({ row: n, message: "Identité répétée : vérifier avant import" });
    seen.add(identity);
    if (row.nationalities) {
      row.nationalities = String(row.nationalities)
        .split("|")
        .map((x) => x.trim().toUpperCase())
        .filter(Boolean);
    }
    const countries: string[] = Array.isArray(row.nationalities) ? [row.country ?? "", ...row.nationalities] : [];
    if (countries.length < 2 || !countries.every(isCountry)) {
      errors.push({ row: n, message: "Pays et nationalités obligatoires : codes de deux lettres majuscules ; séparer les nationalités par |" });
    }
    rows.push(row);
  }
  return { rows, errors, warnings };
}

// --- Prévisualisations en base : table import_previews ------------------------------------------

type StoredPreview = { preview: ImportPreview; version: number };

// Enregistre l'aperçu pour l'utilisateur et purge ceux de plus d'une heure.
export async function savePreview(store: Store, userId: string, preview: ImportPreview): Promise<string> {
  const now = store.clock();
  const version = (await store.read()).version;
  const id = uid();
  const stored: StoredPreview = { preview, version };
  await store.execute("DELETE FROM import_previews WHERE created_at < ?", [now - PREVIEW_SECONDS]);
  await store.execute("INSERT INTO import_previews (id, user_id, created_at, data) VALUES (?, ?, ?, ?)", [id, userId, now, JSON.stringify(stored)]);
  return id;
}

// Applique l'aperçu : mêmes commandes person.save / entry.save que le Python, sous une seule
// transaction, refus si l'état a changé depuis l'aperçu (409). Renvoie le nombre de personnes créées.
export async function commitPreview(store: Store, actor: User, previewId: unknown): Promise<number> {
  const row = typeof previewId === "string" && previewId ? (await store.execute("SELECT user_id, created_at, data FROM import_previews WHERE id = ?", [previewId]))[0] : undefined;
  if (!row || row.user_id !== actor.id || store.clock() - Number(row.created_at) > PREVIEW_SECONDS) throw new Problem("Aperçu expiré ou inconnu.");
  const stored: StoredPreview = JSON.parse(String(row.data));
  if (stored.preview.errors.length) throw new Problem("Corrigez les erreurs avant import.");
  const count = await store.transact(async (tx: Conn) => {
    const s = await store.read(tx);
    if (s.version !== stored.version) throw new Problem("Refaites l’aperçu après les modifications récentes.", 409);
    const before = s.version;
    let count = 0;
    for (const item of stored.preview.rows) {
      const { category_id, ...person } = item;
      person.id = uid();
      await applyCommand(store, tx, s, actor, "person.save", { person });
      if (category_id) await applyCommand(store, tx, s, actor, "entry.save", { entry: { id: uid(), person_id: person.id, category_id, confirmed: false } });
      count += 1;
    }
    s.version += 1;
    await store.write(tx, s, before);
    await store.record(tx, s, actor.id, "import.commit", { count });
    await tx.execute({ sql: "DELETE FROM import_previews WHERE id = ?", args: [previewId as string] });
    return count;
  });
  return count;
}
